packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -973,6 +992,105 @@ type DatabaseRemovedAction = {
   namespace: string;
 };
 
+/**
+ * Converts saved workspace state data back to OpenWorkspaceOptions format
+ * for initializing the store
+ */
+function convertSavedStateToOpenWorkspaceOptions(
+  savedState: WorkspacesStateData
+): OpenWorkspaceOptions[] {
+  return savedState.tabs.map((tab) => {
+    const baseTab: Record<string, unknown> = { type: tab.type };
+
+    // Add connection-related fields
+    if (tab.connectionId) {
+      baseTab.connectionId = tab.connectionId;
+    }
+    if (tab.namespace) {
+      baseTab.namespace = tab.namespace;
+    }
+
+    // Add optional fields based on workspace type
+    if (tab.initialQuery) {
+      baseTab.initialQuery = tab.initialQuery;
+    }
+    if (tab.initialAggregation) {
+      baseTab.initialAggregation = tab.initialAggregation;
+    }
+    if (tab.initialPipeline) {
+      baseTab.initialPipeline = tab.initialPipeline;
+    }
+    if (tab.initialPipelineText) {
+      baseTab.initialPipelineText = tab.initialPipelineText;
+    }
+    if (tab.editViewName) {
+      baseTab.editViewName = tab.editViewName;
+    }
+    if (tab.initialEvaluate) {
+      baseTab.initialEvaluate = tab.initialEvaluate;
+    }
+    if (tab.initialInput) {
+      baseTab.initialInput = tab.initialInput;
+    }
+    if (tab.subTab) {
+      baseTab.initialSubtab = tab.subTab;
+    }
+
+    return baseTab as OpenWorkspaceOptions;
+  });
+}
+
+export const loadSavedWorkspaces = (): WorkspacesThunkAction<
+  Promise<void>,
+  RestoreWorkspacesAction
+> => {
+  return async (dispatch, getState, { connections, userData }) => {
```

Comment:
I think we should move this feature behind a feature flag, it's a pretty big change to compass behavior and we'd want to test it first in dev and beta before releasing. I would also allow you to move forward with merging this even as there are still follow-ups to do, avoiding this branch going too stale (that's our usual approach to developing features substantial in size)

```suggestion
  return async (dispatch, getState, { connections, userData, preferences }) => {
    if (!preferences.getPreferences().enableRestoreWorkspaces) {
      return;
    }
```